src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,298 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import logging
+import os
+from collections.abc import Iterable
+from pathlib import Path
+from typing import Any, Callable, Literal, TypeVar
+
+import tomlkit
+from tomlkit import TOMLDocument
+from tomlkit.items import Table
+
+from ..errors import ConfigParserError, ConfigSourceError
+
+_T = TypeVar("_T")
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ConfigOption:
+    """ConfigOption represents a flag/setting.
+
+    The class knows how to read the value out of all sources and implements
+    order of precedence between them.
+
+    Attributes:
+        name: Name of this ConfigOption.
+        parse_str: A function that can turn str to the desired type, useful
+          for reading value from environmental variable.
+        choices: An iterable of all possible values that are allowed for
+          this option.
+        env_name: Environmental variable value should be read from, if not
+          supplied, we'll construct this. False disables reading from
+          environmental variable, None uses the auto generated variable name
+          and explicitly provided string overwrites the default one.
+        _root_parser: Reference to the root parser. Used to efficiently
+          refer to cached config file. Is supplied by the parent
+          ConfigParser.
+        _nest_path: The names of the ConfigParsers that this option is
+          nested in. Used be able to efficiently resolve where to grab
+          value out of configuration file and construct environment
+          variable name. This is supplied by the parent ConfigParser.
+    """
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        _root_parser: ConfigParser,
+        _nest_path: list[str],
+        parse_str: Callable[[str], _T] | None = None,
+        choices: Iterable[Any] | None = None,
+        env_name: str | None | Literal[False] = None,
+    ) -> None:
+        """Create a config option that can read values from different locations.
+
+        Args:
+            name: Name to assign to this ConfigOption.
+            parse_str: String parser function for this ConfigOption.
+            choices: List of possible values for this ConfigOption.
+            env_name: Environmental variable name value should be read from.
+            _root_parser: Reference to the root parser. Should be supplied by
+              the parent ConfigParser.
+            _nest_path: The names of the ConfigParsers that this option is
+              nested in. This is supplied by the parent ConfigParser.
+        """
+        self.name = name
+        self.parse_str = parse_str
+        self.choices = choices
+        self._nest_path = _nest_path + [name]
+        self._root_parser: ConfigParser = _root_parser
+        self.env_name = env_name
+
+    def get(self) -> Any:
```

Comment:
Nit: If this is an option, why not maybe follow the .get() function layout for dictionaries where a default value can be added when a key is not found?

If such behavior isn't desired, why not use a property `.value` or so? This would be similar to `option_name` below.

Reply:
Sure, `value` it is then!